helper/terraformtype/helper/schema/type_resource.go

```diff
@@ -33,6 +33,7 @@ type ResourceInfo struct {
 	Fields          map[string]*ast.KeyValueExpr
 	Resource        *tfschema.Resource
 	TypesInfo       *types.Info
+	SchemaNames     []string
```

Comment:
I wan to keep the order of the schema definitions, so that some other analyzer could make use of that (as the order is lost in `map`), e.g. check the schema definition order conform to some rules.

Reply:
The AST automatically preserves ordering of source code in `*ast.CompositeLit` `Elts`, e.g.

```go
package main

import (
	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
)

func main() {
	_ = map[string]*schema.Schema{
		"name3": &schema.Schema{},
		"name1": &schema.Schema{},
		"name2": &schema.Schema{},
	}
}
```

Dumps as:

```
    53  .  .  .  .  .  .  .  0: *ast.CompositeLit {
    54  .  .  .  .  .  .  .  .  Type: *ast.MapType {
    55  .  .  .  .  .  .  .  .  .  Map: main.go:8:6
    56  .  .  .  .  .  .  .  .  .  Key: *ast.Ident {
    57  .  .  .  .  .  .  .  .  .  .  NamePos: main.go:8:10
    58  .  .  .  .  .  .  .  .  .  .  Name: "string"
    59  .  .  .  .  .  .  .  .  .  }
    60  .  .  .  .  .  .  .  .  .  Value: *ast.StarExpr {
    61  .  .  .  .  .  .  .  .  .  .  Star: main.go:8:17
    62  .  .  .  .  .  .  .  .  .  .  X: *ast.SelectorExpr {
    63  .  .  .  .  .  .  .  .  .  .  .  X: *ast.Ident {
    64  .  .  .  .  .  .  .  .  .  .  .  .  NamePos: main.go:8:18
    65  .  .  .  .  .  .  .  .  .  .  .  .  Name: "schema"
    66  .  .  .  .  .  .  .  .  .  .  .  }
    67  .  .  .  .  .  .  .  .  .  .  .  Sel: *ast.Ident {
    68  .  .  .  .  .  .  .  .  .  .  .  .  NamePos: main.go:8:25
    69  .  .  .  .  .  .  .  .  .  .  .  .  Name: "Schema"
    70  .  .  .  .  .  .  .  .  .  .  .  }
    71  .  .  .  .  .  .  .  .  .  .  }
    72  .  .  .  .  .  .  .  .  .  }
    73  .  .  .  .  .  .  .  .  }
    74  .  .  .  .  .  .  .  .  Lbrace: main.go:8:31
    75  .  .  .  .  .  .  .  .  Elts: []ast.Expr (len = 3) {
    76  .  .  .  .  .  .  .  .  .  0: *ast.KeyValueExpr {
    77  .  .  .  .  .  .  .  .  .  .  Key: *ast.BasicLit {
    78  .  .  .  .  .  .  .  .  .  .  .  ValuePos: main.go:9:3
    79  .  .  .  .  .  .  .  .  .  .  .  Kind: STRING
    80  .  .  .  .  .  .  .  .  .  .  .  Value: "\"name3\""
    81  .  .  .  .  .  .  .  .  .  .  }
    82  .  .  .  .  .  .  .  .  .  .  Colon: main.go:9:10
    83  .  .  .  .  .  .  .  .  .  .  Value: *ast.UnaryExpr {
    84  .  .  .  .  .  .  .  .  .  .  .  OpPos: main.go:9:12
    85  .  .  .  .  .  .  .  .  .  .  .  Op: &
    86  .  .  .  .  .  .  .  .  .  .  .  X: *ast.CompositeLit {
    87  .  .  .  .  .  .  .  .  .  .  .  .  Type: *ast.SelectorExpr {
    88  .  .  .  .  .  .  .  .  .  .  .  .  .  X: *ast.Ident {
    89  .  .  .  .  .  .  .  .  .  .  .  .  .  .  NamePos: main.go:9:13
    90  .  .  .  .  .  .  .  .  .  .  .  .  .  .  Name: "schema"
    91  .  .  .  .  .  .  .  .  .  .  .  .  .  }
    92  .  .  .  .  .  .  .  .  .  .  .  .  .  Sel: *ast.Ident {
    93  .  .  .  .  .  .  .  .  .  .  .  .  .  .  NamePos: main.go:9:20
    94  .  .  .  .  .  .  .  .  .  .  .  .  .  .  Name: "Schema"
    95  .  .  .  .  .  .  .  .  .  .  .  .  .  }
    96  .  .  .  .  .  .  .  .  .  .  .  .  }
    97  .  .  .  .  .  .  .  .  .  .  .  .  Lbrace: main.go:9:26
    98  .  .  .  .  .  .  .  .  .  .  .  .  Rbrace: main.go:9:27
    99  .  .  .  .  .  .  .  .  .  .  .  .  Incomplete: false
   100  .  .  .  .  .  .  .  .  .  .  .  }
   101  .  .  .  .  .  .  .  .  .  .  }
   102  .  .  .  .  .  .  .  .  .  }
   103  .  .  .  .  .  .  .  .  .  1: *ast.KeyValueExpr {
   104  .  .  .  .  .  .  .  .  .  .  Key: *ast.BasicLit {
   105  .  .  .  .  .  .  .  .  .  .  .  ValuePos: main.go:10:3
   106  .  .  .  .  .  .  .  .  .  .  .  Kind: STRING
   107  .  .  .  .  .  .  .  .  .  .  .  Value: "\"name1\""
   108  .  .  .  .  .  .  .  .  .  .  }
   109  .  .  .  .  .  .  .  .  .  .  Colon: main.go:10:10
   110  .  .  .  .  .  .  .  .  .  .  Value: *ast.UnaryExpr {
   111  .  .  .  .  .  .  .  .  .  .  .  OpPos: main.go:10:12
   112  .  .  .  .  .  .  .  .  .  .  .  Op: &
   113  .  .  .  .  .  .  .  .  .  .  .  X: *ast.CompositeLit {
   114  .  .  .  .  .  .  .  .  .  .  .  .  Type: *ast.SelectorExpr {
   115  .  .  .  .  .  .  .  .  .  .  .  .  .  X: *ast.Ident {
   116  .  .  .  .  .  .  .  .  .  .  .  .  .  .  NamePos: main.go:10:13
   117  .  .  .  .  .  .  .  .  .  .  .  .  .  .  Name: "schema"
   118  .  .  .  .  .  .  .  .  .  .  .  .  .  }
   119  .  .  .  .  .  .  .  .  .  .  .  .  .  Sel: *ast.Ident {
   120  .  .  .  .  .  .  .  .  .  .  .  .  .  .  NamePos: main.go:10:20
   121  .  .  .  .  .  .  .  .  .  .  .  .  .  .  Name: "Schema"
   122  .  .  .  .  .  .  .  .  .  .  .  .  .  }
   123  .  .  .  .  .  .  .  .  .  .  .  .  }
   124  .  .  .  .  .  .  .  .  .  .  .  .  Lbrace: main.go:10:26
   125  .  .  .  .  .  .  .  .  .  .  .  .  Rbrace: main.go:10:27
   126  .  .  .  .  .  .  .  .  .  .  .  .  Incomplete: false
   127  .  .  .  .  .  .  .  .  .  .  .  }
   128  .  .  .  .  .  .  .  .  .  .  }
   129  .  .  .  .  .  .  .  .  .  }
   130  .  .  .  .  .  .  .  .  .  2: *ast.KeyValueExpr {
   131  .  .  .  .  .  .  .  .  .  .  Key: *ast.BasicLit {
   132  .  .  .  .  .  .  .  .  .  .  .  ValuePos: main.go:11:3
   133  .  .  .  .  .  .  .  .  .  .  .  Kind: STRING
   134  .  .  .  .  .  .  .  .  .  .  .  Value: "\"name2\""
   135  .  .  .  .  .  .  .  .  .  .  }
   136  .  .  .  .  .  .  .  .  .  .  Colon: main.go:11:10
   137  .  .  .  .  .  .  .  .  .  .  Value: *ast.UnaryExpr {
   138  .  .  .  .  .  .  .  .  .  .  .  OpPos: main.go:11:12
   139  .  .  .  .  .  .  .  .  .  .  .  Op: &
   140  .  .  .  .  .  .  .  .  .  .  .  X: *ast.CompositeLit {
   141  .  .  .  .  .  .  .  .  .  .  .  .  Type: *ast.SelectorExpr {
   142  .  .  .  .  .  .  .  .  .  .  .  .  .  X: *ast.Ident {
   143  .  .  .  .  .  .  .  .  .  .  .  .  .  .  NamePos: main.go:11:13
   144  .  .  .  .  .  .  .  .  .  .  .  .  .  .  Name: "schema"
   145  .  .  .  .  .  .  .  .  .  .  .  .  .  }
   146  .  .  .  .  .  .  .  .  .  .  .  .  .  Sel: *ast.Ident {
   147  .  .  .  .  .  .  .  .  .  .  .  .  .  .  NamePos: main.go:11:20
   148  .  .  .  .  .  .  .  .  .  .  .  .  .  .  Name: "Schema"
   149  .  .  .  .  .  .  .  .  .  .  .  .  .  }
   150  .  .  .  .  .  .  .  .  .  .  .  .  }
   151  .  .  .  .  .  .  .  .  .  .  .  .  Lbrace: main.go:11:26
   152  .  .  .  .  .  .  .  .  .  .  .  .  Rbrace: main.go:11:27
   153  .  .  .  .  .  .  .  .  .  .  .  .  Incomplete: false
   154  .  .  .  .  .  .  .  .  .  .  .  }
   155  .  .  .  .  .  .  .  .  .  .  }
   156  .  .  .  .  .  .  .  .  .  }
   157  .  .  .  .  .  .  .  .  }
   158  .  .  .  .  .  .  .  .  Rbrace: main.go:12:2
   159  .  .  .  .  .  .  .  .  Incomplete: false
   160  .  .  .  .  .  .  .  }
```

Storing the list of strings would be lossy as well since we'd lose the `ast.BasicLit` `token.Pos`.

You may find that the `helper/terraformtype/helper/schema.GetSchemaMapAttributeNames()` helper function after analyzing via `passes/helper/schema/schemamapcompositelit` gives the information I think you are looking for. See S015 for an example of analyzing attribute names.

Reply:
Thank you for pointing this out, I will remove that field then.

Reply:
Thank you for pointing this out, I will remove that field then.